tests/service_controller_test.go

```diff
@@ -0,0 +1,75 @@
+package tests
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"kubevirt.io/ssp-operator/controllers"
+	"kubevirt.io/ssp-operator/internal/operands/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func getSspMetricsService() (*v1.Service, error) {
+	service := controllers.ServiceObject(strategy.GetSSPDeploymentNameSpace())
+	err := apiClient.Get(ctx, client.ObjectKeyFromObject(service), service)
+	return service, err
+}
+
+func equalService(serviceA, serviceB *v1.Service) bool {
+	return reflect.DeepEqual(serviceA.Labels, serviceB.Labels) && reflect.DeepEqual(serviceA.Spec, serviceB.Spec)
+}
+
+var _ = Describe("Service Controller", func() {
+	var (
+		serviceErr error
+		service    *v1.Service
+	)
+	BeforeEach(func() {
+		waitUntilDeployed()
+	})
+
+	It("[test_id: 8807] Should create ssp-operator-metrics service", func() {
+		service, serviceErr = getSspMetricsService()
+		if serviceErr != nil {
+			Fail("Failed to get ssp-operator-metrics service")
+			return
+		}
+	})
+
+	It("[test_id: 8808] Should re-create ssp-operator-metrics service if deleted", func() {
+		service, serviceErr = getSspMetricsService()
+		oldUID := service.UID
+		Expect(apiClient.Delete(ctx, service)).To(Succeed())
+		Eventually(func() error {
+			err := apiClient.Get(ctx, client.ObjectKeyFromObject(service), service)
+			return err
+		}, shortTimeout, time.Second).Should(Succeed(), fmt.Sprintf("Did not recreate the %s service", controllers.MetricsServiceName))
+		Expect(service.UID).NotTo(Equal(oldUID))
+	})
+
+	It("[test_id: 8810] Should restore ssp-operator-metrics service after update", func() {
+		service, serviceErr = getSspMetricsService()
+		changed := service.DeepCopyObject().(*v1.Service)
+		changed.Labels = nil
+		changed.Spec.Ports = []v1.ServicePort{
+			{
+				Name:       metrics.MetricsPortName,
+				Port:       755,
+				Protocol:   v1.ProtocolTCP,
+				TargetPort: intstr.FromString(metrics.MetricsPortName),
+			},
+		}
+
+		Expect(apiClient.Update(ctx, changed)).ToNot(HaveOccurred())
```

Comment:
The `Update` needs to be wrapped in an `Eventually()` because it may fail when the object is updated concurrently.
```go
Eventually(func() error {
   return apiClient.Update(ctx, changed)
}, shortTimeout, time.Second).Should(Succeed())
```

Reply:
ah you are right!